[TECH] Refacto getNextChallengeForCertification (PIX-10870)

mcampourcy opened the pull request:
## :unicorn: Problème
La méthode getNextChallengeForCertification est très lourde, et contient des spécificités v2/v3

## :robot: Proposition
Séparer en 2 fichiers distincts ce qui touche à v2 et ce qui touche à v3

## :rainbow: Remarques
Il n'y a eu aucune réécriture de la méthode proprement dite, juste une création de 2 fichiers : `get-next-challenge-for-v2-certification.js` et `get-next-challenge-for-v3-certification.js`, en lieu et place de `get-next-challenge-for-certification.js`

## :100: Pour tester

- Créer une session de certif v3 [certifv3@example.net](mailto:certifv3@example.net) // pix123
- Rejoindre la session de certif v3 [certifiable-contenu-user@example.net](mailto:certifiable-contenu-user@example.net) // pix123
- Les questions s'affichent correctement jusqu'à la fin
---
- Créer une session de certif v2 [certif-pro@example.net](mailto:certifv3@example.net) // pix123
- Rejoindre la session de certif v2 [certifiable-contenu-user@example.net](mailto:certifiable-contenu-user@example.net) // pix123
- Les questions s'affichent correctement jusqu'à la fin

mcampourcy committed:
:recycle: api: Refacto getNextChallengeForCertification to split it for v2/v3

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr8231.review.pix.fr)
  * [App (.org)](https://app-pr8231.review.pix.org)
  * [Orga (.fr)](https://orga-pr8231.review.pix.fr)
  * [Orga (.org)](https://orga-pr8231.review.pix.org)
  * [Certif (.fr)](https://certif-pr8231.review.pix.fr)
  * [Certif (.org)](https://certif-pr8231.review.pix.org)
  * [1D](https://1d-pr8231.review.pix.fr)
  * [Admin](https://admin-pr8231.review.pix.fr)
  * [API](https://api-pr8231.review.pix.fr/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr8231.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr8231/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr8231/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr8231/environment)

aceol commented on api/lib/domain/usecases/index.js:
```diff
@@ -392,7 +393,8 @@ const usecasesWithoutInjectedDependencies = {
   findPaginatedFilteredTargetProfileOrganizations,
   getCampaignDetailsManagement,
-  getNextChallengeForCertification,
+  getNextChallengeForV2Certification: getNextChallengeForV2Certification,
+  getNextChallengeForV3Certification: getNextChallengeForV3Certification,
```
```suggestion
  getNextChallengeForV2Certification,
  getNextChallengeForV3Certification,
```

aceol commented on api/src/certification/course/domain/usecases/get-next-challenge-for-v2-certification.js:
```diff
@@ -0,0 +1,13 @@
+const getNextChallengeForV2Certification = async function ({
```
Ce n'est plus d'actualité, mais il y a une v1!
On a besoin du prochain challenge, dans un cas il est deja attribué, dans l'autre il l'est fait dynamiquement
on pourrait avoir par exemple: `getNextPreSelectedChallenge` et `getNextSelectedChallenge` 
?

alexandrecoin replied:
`getNextChallengeForPreviousCertificationVersions` ?

aceol replied:
C'est vrai qu'on va pas recuperer les challenges sur une V1...

aceol commented on api/src/certification/course/domain/usecases/get-next-challenge-for-v2-certification.js:
```diff
@@ -0,0 +1,13 @@
+  certificationChallengeRepository,
+  challengeRepository,
+}) {
+  return certificationChallengeRepository
+    .getNextNonAnsweredChallengeByCourseId(assessment.id, assessment.certificationCourseId)
+    .then((certificationChallenge) => {
+      return challengeRepository.get(certificationChallenge.challengeId);
+    });
```
On sort un peu du cadre mais: 

```suggestion
  const certificationChallenge = await certificationChallengeRepository
    .getNextNonAnsweredChallengeByCourseId(assessment.id, assessment.certificationCourseId)
  return challengeRepository.get(certificationChallenge.challengeId);
```

aceol commented on api/src/shared/application/assessments/assessment-controller.js:
```diff
@@ -194,7 +197,13 @@ async function _getChallengeByAssessmentType({ assessment, request, dependencies
-    return dependencies.usecases.getNextChallengeForCertification({ assessment, locale });
+    const certificationCourse = await dependencies.certificationCourseRepository.get(assessment.certificationCourseId);
+
+    if (certificationCourse.getVersion() === CertificationVersion.V3) {
```
```suggestion
    if (certificationCourse.isV3()) {
```

alexandrecoin replied:
Pour rebondir également sur le commentaire lié à l'optimisation, je propose d'ajouter une méthode de repo pour récupérer uniquement un certif course via l'id et d'en extraire la version via décomposition dans le controller.

```
const { version : certificationVersion } = await dependencies.certificationCourseRepository.getBareCertificationCourse(assessment.certificationCourseId)
```

Afin de ne pas utiliser la méthode de l'objet du domaine `isV3()` qui n'a à mon avis rien à faire dans le controleur.

aceol replied:
Ca n'aiderais pas à l'optimisation
> Ici on va chercher le certif course, ses challenges, son assessment, les certif-reports pour uniquement savoir si on est en v3

aceol commented on api/src/shared/application/assessments/assessment-controller.js:
```diff
@@ -194,7 +197,13 @@ async function _getChallengeByAssessmentType({ assessment, request, dependencies
 
   if (assessment.isCertification()) {
-    return dependencies.usecases.getNextChallengeForCertification({ assessment, locale });
+    const certificationCourse = await dependencies.certificationCourseRepository.get(assessment.certificationCourseId);
```
Ici on va chercher le certif course, ses challenges, son assessment, les certif-reports pour uniquement savoir si on est en v3. On pourrait bcp optimiser

aceol reviewed (commented):
A voir si on peut renommer (pour ne pas "oublier" la v1)
et aussi optimiser pour ne recuperer que la version de la certification

Commit:
:recycle: api: asyncify usecase

Commit:
:sparkles: api: add certification version repository

Commit:
:recycle: api: reorder imports due to linter configuration modification

This renaming allows us to take into account the fact BOTH V1 and V2 certifications
share the same process of challenge selection

aceol commented on api/src/certification/course/infrastructure/repositories/certification-version-repository.js:
```diff
@@ -0,0 +1,23 @@
+};
+
+const _toDomain = (certificationVersion) => {
+  const key = _getKeyByValue(CertificationVersion, certificationVersion);
```
Petit noeuds au cerveau, j'ai l'impression qu'on fait bcp pour pas grand chose...

alexandrecoin replied:
Pour ne pas dire à rien, c'est enlevé :D

aceol commented on api/src/shared/application/assessments/assessment-controller.js:
```diff
@@ -199,9 +201,11 @@ async function _getChallengeByAssessmentType({ assessment, request, dependencies
+    });
 
-    if (certificationCourse.getVersion() === CertificationVersion.V3) {
+    if (certificationCourseVersion === CertificationVersion.V3) {
```
Sur un certificationVersion, on pourrait pas avoir un petit isV3?

alexandrecoin replied:
Cette modification va changer pas mal de fichiers, on sort un peu trop du scope de cette PR.

aceol commented on api/src/certification/course/domain/usecases/get-next-challenge-for-non-v3-certification.js:
```diff
@@ -1,4 +1,4 @@
-const getNextChallengeForV2Certification = async function ({
+const getNextChallengeForNonV3Certification = async function ({
```
et si y'a une v4? 
Ce nom me semble pas judicieux, je prefere le v2...mince
